File: check_strings.py
#!/usr/bin/env python3
"""check_strings.py — equivalent exact de check-strings.js, pour les postes
sans Node. Meme verdict, meme code de sortie.

    python check_strings.py

La specification demande check-strings.js ; ce jumeau existe parce que le
poste de deploiement de ce projet a Python mais pas Node, et qu'une barriere
qu'on ne peut pas executer ne protege rien.
"""
import json
import os
import re
import sys

ROOT = os.path.dirname(os.path.abspath(__file__))
LANGS = ("fr", "en")
errors = []
warnings = []

# --- 1. lecture -----------------------------------------------------------

with open(os.path.join(ROOT, "strings.json"), "rb") as f:
    raw = f.read()

if raw[:3] == b"\xef\xbb\xbf":
    errors.append("strings.json commence par un BOM — il doit etre en UTF-8 sans BOM.")

try:
    dico = json.loads(raw.decode("utf-8"))
except Exception as e:                                    # noqa: BLE001
    print("strings.json illisible : %s" % e, file=sys.stderr)
    sys.exit(1)

keys = [k for k in dico if k != "_comment"]

# --- 2. completude des deux langues ---------------------------------------

for k in keys:
    entry = dico[k]
    if not isinstance(entry, dict):
        errors.append("« %s » n'est pas un objet { fr, en }." % k)
        continue
    for l in LANGS:
        if l not in entry:
            errors.append("« %s » n'a pas de traduction %s." % (k, l))
        elif not isinstance(entry[l], str) or not entry[l].strip():
            errors.append("« %s » a une traduction %s vide." % (k, l))

# --- 3. coherence des marqueurs {xxx} -------------------------------------

PH = re.compile(r"\{[A-Za-z0-9_]+\}")


def placeholders(s):
    return ",".join(sorted(PH.findall(str(s))))


for k in keys:
    e = dico[k]
    if not isinstance(e, dict):
        continue
    if not all(isinstance(e.get(l), str) for l in LANGS):
        continue
    a, b = placeholders(e["fr"]), placeholders(e["en"])
    if a != b:
        errors.append(
            "« %s » : marqueurs differents entre les langues — fr [%s] contre en [%s]." % (k, a, b)
        )

# --- 4. cles referencees par le code --------------------------------------

namespaces = {k.split(".")[0] for k in keys}
used = set()

with open(os.path.join(ROOT, "index.html"), encoding="utf-8") as f:
    html = f.read()
for m in re.finditer(r'data-i18n(?:-ph)?="([^"]+)"', html):
    used.add(m.group(1))

with open(os.path.join(ROOT, "app.js"), encoding="utf-8") as f:
    js = f.read()
for m in re.finditer(r"'([a-z0-9_]+\.[a-z0-9_]+)'", js):
    cand = m.group(1)
    if cand.split(".")[0] in namespaces:
        used.add(cand)
    else:
        warnings.append("litteral « %s » ignore : espace de noms inconnu." % cand)

for k in sorted(used):
    if k not in keys:
        errors.append("cle « %s » utilisee par le code mais absente de strings.json." % k)

# --- 5. cles jamais utilisees : signale, pas bloquant ---------------------

for k in keys:
    if k not in used:
        warnings.append("cle « %s » definie mais jamais utilisee." % k)

# --- 6. verdict ------------------------------------------------------------

print("%d cles · %d referencees par le code" % (len(keys), len(used)))

if warnings:
    print("\n%d avertissement(s) :" % len(warnings))
    for w in warnings:
        print("  · %s" % w)

if errors:
    print("\n%d ERREUR(S) — deploiement interdit :" % len(errors), file=sys.stderr)
    for e in errors:
        print("  x %s" % e, file=sys.stderr)
    sys.exit(1)

print("\nOK — les deux langues sont completes et coherentes.")
sys.exit(0)

File: check-strings.js
#!/usr/bin/env node
/* check-strings.js — barrière avant déploiement (§3).
 *
 *   node check-strings.js
 *
 * Sort avec le code 1 si :
 *   - une clé manque en français ou en anglais, ou est vide ;
 *   - une clé référencée par index.html ou app.js n'existe pas ;
 *   - strings.json porte un BOM (il est lu par fetch, pas par Excel) ;
 *   - les marqueurs {xxx} d'une langue ne sont pas ceux de l'autre.
 *
 * Ce dernier point attrape l'erreur discrète : traduire « {n} arbres » par
 * « trees planned » sans le marqueur produit un écran vide de son chiffre,
 * sans lever la moindre exception au champ.
 *
 * Un équivalent Python — check_strings.py — est livré à côté : il donne
 * exactement le même verdict, pour les postes sans Node.
 */
'use strict';

const fs = require('fs');
const path = require('path');

const ROOT = __dirname;
const LANGS = ['fr', 'en', 'ru'];
const errors = [];
const warnings = [];

/* --- 1. lecture ------------------------------------------------------- */

const raw = fs.readFileSync(path.join(ROOT, 'strings.json'));
if (raw[0] === 0xEF && raw[1] === 0xBB && raw[2] === 0xBF) {
  errors.push('strings.json commence par un BOM — il doit être en UTF-8 sans BOM.');
}

let dict;
try {
  dict = JSON.parse(raw.toString('utf8'));
} catch (e) {
  console.error('strings.json illisible : ' + e.message);
  process.exit(1);
}

const keys = Object.keys(dict).filter(k => k !== '_comment');

/* --- 2. complétude des deux langues ----------------------------------- */

for (const k of keys) {
  const entry = dict[k];
  if (typeof entry !== 'object' || entry === null) {
    errors.push('« ' + k + ' » n\'est pas un objet { fr, en }.');
    continue;
  }
  for (const l of LANGS) {
    if (!(l in entry)) errors.push('« ' + k + ' » n\'a pas de traduction ' + l + '.');
    else if (typeof entry[l] !== 'string' || entry[l].trim() === '') {
      errors.push('« ' + k +' » a une traduction ' + l + ' vide.');
    }
  }
}

/* --- 3. cohérence des marqueurs {xxx} --------------------------------- */

const placeholders = (s) => (String(s).match(/\{[a-z0-9_]+\}/gi) || []).sort().join(',');

/* Le français fait référence : chaque langue doit porter exactement les mêmes
   marqueurs. Une traduction qui perd son {n} affiche un écran privé de son
   chiffre, sans lever la moindre exception au champ. */
for (const k of keys) {
  const e = dict[k];
  if (!e || typeof e !== 'object') continue;
  if (!LANGS.every(l => typeof e[l] === 'string')) continue;
  const ref = placeholders(e.fr);
  for (const l of LANGS) {
    if (l === 'fr') continue;
    const got = placeholders(e[l]);
    if (got !== ref) {
      errors.push('« ' + k + ' » : marqueurs différents — fr [' + ref
        + '] contre ' + l + ' [' + got + '].');
    }
  }
}

/* --- 4. clés référencées par le code ----------------------------------- */

const namespaces = new Set(keys.map(k => k.split('.')[0]));
const used = new Set();

const html = fs.readFileSync(path.join(ROOT, 'index.html'), 'utf8');
for (const m of html.matchAll(/data-i18n(?:-ph)?="([^"]+)"/g)) used.add(m[1]);

/* Dans app.js les clés arrivent aussi comme arguments littéraux
   (bigButton, toast, dataMsg, figure…), pas seulement dans t('…').
   On relève donc tout littéral ayant la forme d'une clé, et on ne retient
   que ceux dont l'espace de noms existe — « strings.json » ou « app.js »
   ont la même forme mais ne sont pas des clés. */
const js = fs.readFileSync(path.join(ROOT, 'app.js'), 'utf8');
for (const m of js.matchAll(/'([a-z0-9_]+\.[a-z0-9_]+)'/g)) {
  const cand = m[1];
  if (namespaces.has(cand.split('.')[0])) used.add(cand);
  else warnings.push('littéral « ' + cand + ' » ignoré : espace de noms inconnu.');
}

for (const k of used) {
  if (!keys.includes(k)) errors.push('clé « ' + k + ' » utilisée par le code mais absente de strings.json.');
}

/* --- 5. clés jamais utilisées : signalé, pas bloquant ------------------ */

for (const k of keys) {
  if (!used.has(k)) warnings.push('clé « ' + k + ' » définie mais jamais utilisée.');
}

/* --- 6. verdict -------------------------------------------------------- */

console.log(keys.length + ' clés · ' + used.size + ' référencées par le code');

if (warnings.length) {
  console.log('\n' + warnings.length + ' avertissement(s) :');
  warnings.forEach(w => console.log('  · ' + w));
}

if (errors.length) {
  console.error('\n' + errors.length + ' ERREUR(S) — déploiement interdit :');
  errors.forEach(e => console.error('  ✗ ' + e));
  process.exit(1);
}

console.log('\nOK — les ' + LANGS.length + ' langues sont complètes et cohérentes.');
process.exit(0);
